Packs/Slack/Integrations/SlackV3/SlackV3.py

```diff
@@ -129,47 +259,30 @@ def get_user_by_name(user_to_search: str, add_to_context: bool = True) -> dict:
-                                             or u.get('profile', {}).get('email', '').lower() == user_to_search
-                                             or u.get('real_name', '').lower() == user_to_search, users))
-        if users_filter:
-            user = users_filter[0]
-    if not user:
-        body = {
-            'limit': PAGINATED_COUNT
-        }
-
-        response = send_slack_request_sync(CLIENT, 'users.list', http_verb='GET', body=body)
-        while True:
-            workspace_users = response['members'] if response and response.get('members',
-                                                                               []) else []
-            cursor = response.get('response_metadata', {}).get('next_cursor')
-            users_filter = list(filter(lambda u: u.get('name', '').lower() == user_to_search
-                                       or u.get('profile', {}).get('email', '').lower() == user_to_search
-                                       or u.get('real_name', '').lower() == user_to_search, workspace_users))
-            if users_filter:
-                break
-            if not cursor:
-                break
-            body = body.copy()  # strictly for unit testing purposes
-            body.update({'cursor': cursor})
-            response = send_slack_request_sync(CLIENT, 'users.list', http_verb='GET', body=body)
-
-        if users_filter:
-            user = users_filter[0]
-            if add_to_context:
-                users.append(user)
-                set_to_integration_context_with_retries({'users': users}, OBJECTS_TO_KEYS, SYNC_CONTEXT)
-        else:
-            return {}
+    if not SAFE_MODE:
+        integration_context = get_integration_context(SYNC_CONTEXT)
+        if integration_context.get('users'):
+            demisto.debug(f"Checking in context for {user_to_search}")
+            # Check if we already have the user to prevent call to users.lookupByEmail
+            users = json.loads(integration_context['users'])
+            user = return_user_filter(user_to_search, users)
+
+    if not user and re.match(emailRegex, user_to_search):
+        demisto.debug(f"Checking via API for email of {user_to_search}")
+        user = get_user_by_email(user_to_search)
+        if user and (add_to_context or not SAFE_MODE):
+            integration_context = get_integration_context(SYNC_CONTEXT)
+            add_user_to_context(user=user, integration_context=integration_context)
+    if not user and not SAFE_MODE:
+        demisto.debug(f"Couldn't find {user_to_search} and safe mode is disabled. Checking API")
+        user = paginated_search_for_user(user_to_search)
+        demisto.debug(f"Found {user_to_search} - {user}")
+        if user and (add_to_context or not SAFE_MODE):
+            integration_context = get_integration_context(SYNC_CONTEXT)
+            add_user_to_context(user=user, integration_context=integration_context)
```

Comment:
you are checking safe mode too many times here. Think we should consider to split this to a safe mode part and un safe mode part. even if you will have some code repeated. you should have one `If` if it is safe mode.
and the should be covered by a test.